src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -43,34 +54,64 @@ public DependencyGraphSpec(bool isReadOnly)
             _isReadOnly = isReadOnly;
         }
 
+        private bool _refreshRestoreList = false;
+        private IReadOnlyList<string>? _restoreList = null;
+
         /// <summary>
         /// Projects to restore.
         /// </summary>
-        public IReadOnlyList<string> Restore => _restore.ToList();
+        public IReadOnlyList<string> Restore
+        {
+            get
+            {
+                if (_restoreList == null || _refreshRestoreList)
+                {
+                    _restoreList = _restore.ToList();
+                    _refreshRestoreList = false;
+                }
+                return _restoreList;
+            }
+        }
+
+        private bool _refreshProjectsList = false;
+        private IReadOnlyList<PackageSpec>? _restoreProjectList = null;
 
         /// <summary>
         /// All project specs.
         /// </summary>
-        public IReadOnlyList<PackageSpec> Projects => _projects.Values.ToList();
+        public IReadOnlyList<PackageSpec> Projects
+        {
+            get
+            {
+                if (_restoreProjectList == null || _refreshProjectsList)
+                {
+                    _restoreProjectList = _projects.Values.ToList();
```

Comment:
Same question here as I had for the `Restore` property.

Reply:
This isn't used during the core of restore, but outside. 
All accesses are synchronous.